src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,98 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
+@SuppressWarnings("unchecked")
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double SIZE_MULTIPLIER = 1.5;
+    private T[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = (T[]) new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        grow();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index.");
+        }
+        grow();
+        System.arraycopy(elements, index, elements, index + 1, size - index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        verifyIndex(index);
+        return elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        verifyIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        verifyIndex(index);
+        final T removedElement = elements[index];
+        System.arraycopy(elements, index + 1, elements, index, size - index - 1);
+        elements[--size] = null;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        T removedElement = (T) new Object();
+        for (int i = 0; i < size; i++) {
+            if (elements[i] != null ? elements[i].equals(element) : element == null) {
+                removedElement = elements[i];
+                remove(i);
+                return removedElement;
+            }
+        }
+        throw new NoSuchElementException("No such element");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    public void grow() {
+        if (size >= elements.length) {
+            int newSize = (int) (size * SIZE_MULTIPLIER);
+            T[] tempElements = (T[]) new Object[newSize];
+            System.arraycopy(elements, 0, tempElements, 0, elements.length);
+            elements = tempElements;
+        }
+    }
+
+    public void verifyIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Wrong index.");
```

Comment:
always strive to make your exceptions as informative as possible, include all the relevant details, it will make life and debugging so much easier down the road. Here, size and index would be valuable additions. Take a look at Java's exception:
![image](https://github.com/mate-academy/jv-array-list/assets/49767049/bc7a1d6b-7a63-43bd-a73e-fbd46856b416)
Let's do the same